### ISSUE TITLE
Unsafe buffer usage: false positive

### ISSUE THREAD

**Issue by andrew-aitchison** [llvm/llvm-project#64646]
// Ubuntu clang version 16.0.0 (1~exp5ubuntu3)
// clang++-16 -Wunsafe-buffer-usage -c -o unsafe.o unsafe.cpp

```
int f(const int a[10])
{
    return a[1];
}

```

unsafe.cpp:8:17: warning: 'a' is an unsafe pointer used for buffer access [-Wunsafe-buffer-usage]
int f(const int a[10])
      ~~~~~~~~~~^~~~~
unsafe.cpp:10:12: note: used in buffer access here
    return a[1];
           ^
1 warning generated.

Since the size of array 'a' is known, and known to be fixed, at compile-time, the access is safe *and* the compiler can determine this, so the warning is a false positive.

While I appreciate that `std::span` would be better, there is much code out there that is not ready for C++20.

**Comment by LukasHanel** [comment]
We noticed this as well, for **C** code.
I tried adding `static` like `int f(int a[static 10])`, but that does not fix it.

This can also be hidden behind typedef, like:
````
typedef struct {
    int a[10];
} S;
int f(S s)
{
    return s.a[1];
}
````

I guess the check only really works for C++, what are you supposed to do in C?
It even fails this function:
````
int f(uint8_t *buf, size_t buf_len)
{
    for (size_t i = 0; i < buf_len; i++) {
       buf[i] = 1;
   }
   return 0;
}
````

**Comment by Endilll** [comment]
CC @AaronBallman

**Comment by haoNoQ** [comment]
Thank you Aaron! cc @jkorous-apple  @t-rasmud @malavikasamak as well!

I think Aaron captured our logic perfectly: for the purposes of `-Wunsafe-buffer-usage` this code isn't secure enough. Other than `-Wunsafe-buffer-usage`, there's nothing that prevents you from passing a buffer of size 1 into this function: https://godbolt.org/z/vqajz55hf

Of course you can make the same mistake with `std::span`, like construct a `std::span` with a wrong size. But the idea we're going for with `-Wunsafe-buffer-usage` is deeper than that: the code should not only use a `std::span` here, but it should also carry the span all the way from the allocation site to here, and at the allocation site it should use a safe container that's implicitly convertible to `std::span`, so that there never was a need to specify the size separately.

Which is why, as you can see on godbolt (in the machine-readable `-fdiagnostics-parseable-fixits` part that transcends the command-line fixit UI but can be picked up by an IDE), the automatic fixit attached to the warning actually looks like this:
```diff
-static void foo(int array[12]) {
+static void foo(std::span<int> array) {
     array[1] = 10;
 }

+[[clang::unsafe_buffer_usage]]
+static void foo(int array[12]) {
+  return foo(std::span<int>(array, <# size #>));
+}
```
Basically it unbreaks all call sites by providing a compatibility overload, and then puts an attribute on that overload, that screams back at every caller "please use the other, safe overload". Which encourages `std::span` to be propagated as-is from the caller rather than constructed in place. The warning "crawls" back to the allocation site  this way, one function at a time, until the span gets carried all the way from the allocation site to the access site.

There's also a different way to look at your example. Our argument is that `f()` isn't the ultimate authority when it comes to the size of the buffer `a[]`, so it shouldn't try to pretend that it knows the size, and it's not the function we should trust if we want to make our program bounds-safe. But it's still a good idea to allow the function `f()` to communicate the size of the buffer *it can work with* (be it minimum size or exact size), so that the bounds safety issue could have been caught earlier at runtime, or even avoided entirely by informing developers who call the function about the function's contract. In this sense a good parameter type would be `std::span<int, 12>` - a compile-time-sized span, which is actually a thing in the Standard. Maybe we should consider recommending it automatically as a replacement for such constant array types?

> While I appreciate that `std::span` would be better, there is much code out there that is not ready for C++20.

Absolutely. This is why `-Wunsafe-buffer-usage` is turned off by default; it's not trivial to satisfy and you probably shouldn't enable it when your only goal is "gain a little more safety, almost for free". It's not very useful if you don't have access to bounds-safe containers to transition to. Automatic fixits currently don't support arbitrary safe containers, only standard containiers, so its usefulness outside of C++20 is even further reduced.

Also automatic fixits are still being very actively developed, and there definitely some false positives in the warning. In particular, this:
```c++
void f() {
    int arr[12];
    arr[1] = 10;  // warning: unsafe buffer access [-Wunsafe-buffer-usage]
}
```
is a false positive that we still somehow didn't fix but we absolutely intend to fix. In this case bounds information supplied by the array type is actually precise and the function `f()` is the ultimate authority when it comes to such bounds information.

**Comment by andrew-aitchison** [comment]
Thanks.
I had not realized that the 10 in
   int f(const int a[10])
is ignored. Perhaps a warning to that effect, though I also like the suggestion of recommending replacing it with  std::span<int, 10> 

I would be happy to close this as "not a bug" but for LukasHanel's problem with
int f(int a[static 10])
 in C.

**Comment by jkorous-apple** [comment]
@LukasHanel There are two separate things relevant to your comment.
We haven't taught our analysis to reason about constant sizes and constant indices yet. We are aware of this and plan to address it in the future.

The story we had in mind for C is similar to what we have for C++.
1. Any interested C project can define safe alternatives to pointer arithmetic - e. g. a set of functions that are bounds-safe (either by including a runtime bounds-check or because they are well tested or carefully reviewed or whatever else) and using the opt-out pragma to declare these as safe for `-Wunsafe-buffer-usage` analysis.
2. Then they can use the warning to find all the source locations they should change to the new safe primitives.

Since this programming model is opt-in we see no harm in allowing people to use it in whatever way they want to.
At the same time I don't think we are dead set on allowing `-Wunsafe-buffer-usage` to be used with `-x c`. We don't have any users of this approach and now `-fbounds-safety` is becoming a thing.

**Comment by LukasHanel** [comment]
Ok, good to know. I realize this is a new feature and not yet polished to the end. Since we activate -Weverything, we stumbled upon it, deactivated it quickly and eventually had a look at what good it is for our C code at the moment. Looking forward to an improved version in the future.
